include/graphblas/bsp1d/blas3.hpp

```diff
@@ -195,16 +195,118 @@ namespace grb {
 			phase
 		);
 		if( phase == RESIZE ) {
-			if( collectives<>::allreduce( ret, operators::any_or< RC >() ) != SUCCESS ) {
+			if( collectives<>::allreduce(
+					ret, operators::any_or< RC >()
+				) != SUCCESS
+			) {
 				return PANIC;
-			} else {
-				return SUCCESS;
 			}
+			return SUCCESS;
 		}
 		assert( phase == EXECUTE );
 		return internal::checkGlobalErrorStateOrClear( C, ret );
 	}
 
+	template<
+		Descriptor descr = descriptors::no_operation,
+		typename InputType, typename OutputType,
+		typename RIT_L, typename CIT_L, typename NIT_L,
+		typename RIT_A, typename CIT_A, typename NIT_A
+	>
+	RC tril(
+		Matrix< OutputType, BSP1D, RIT_L, CIT_L, NIT_L > &L,
+		const Matrix< InputType, BSP1D, RIT_A, CIT_A, NIT_A > &A,
+		const long int k,
+		const Phase &phase = Phase::EXECUTE,
+		const typename std::enable_if<
+			!grb::is_object< OutputType >::value &&
+			!grb::is_object< InputType >::value &&
+			std::is_convertible< InputType, OutputType >::value
+		>::type * const = nullptr
+	) {
+#ifdef _DEBUG
+		std::cout << "In grb::tril( BSP1D )\n";
+#endif
+		assert( phase != TRY );
+		const RC ret = tril< descr >(
+			internal::getLocal( L ),
+			internal::getLocal( A ),
+			k,
+			phase
+		);
+
+		return internal::checkGlobalErrorStateOrClear( L, ret );
+	}
+
+	template<
+		Descriptor descr = descriptors::no_operation,
+		typename InputType, typename OutputType,
+		typename RIT_L, typename CIT_L, typename NIT_L,
+		typename RIT_A, typename CIT_A, typename NIT_A
+	>
+	RC tril(
+		Matrix< OutputType, BSP1D, RIT_L, CIT_L, NIT_L > &L,
+		const Matrix< InputType, BSP1D, RIT_A, CIT_A, NIT_A > &A,
+		const Phase &phase = Phase::EXECUTE,
+		const typename std::enable_if<
+			!grb::is_object< OutputType >::value &&
+			!grb::is_object< InputType >::value &&
+			std::is_convertible< InputType, OutputType >::value
+		>::type * const = nullptr
+	) {
+		return tril< descr >( L, A, 0, phase );
+	}
+
+	template<
+		Descriptor descr = descriptors::no_operation,
+		typename InputType, typename OutputType,
+		typename RIT_U, typename CIT_U, typename NIT_U,
+		typename RIT_A, typename CIT_A, typename NIT_A
+	>
+	RC triu(
+		Matrix< OutputType, BSP1D, RIT_U, CIT_U, NIT_U > &U,
+		const Matrix< InputType, BSP1D, RIT_A, CIT_A, NIT_A > &A,
+		const long int k,
+		const Phase &phase = Phase::EXECUTE,
+		const typename std::enable_if<
+			!grb::is_object< OutputType >::value &&
+			!grb::is_object< InputType >::value &&
+			std::is_convertible< InputType, OutputType >::value
+		>::type * const = nullptr
+	) {
+#ifdef _DEBUG
+		std::cout << "In grb::triu( BSP1D )\n";
+#endif
+		assert( phase != TRY );
+		const RC ret = triu< descr >(
+			internal::getLocal( U ),
+			internal::getLocal( A ),
+			k,
+			phase
+		);
+
+		return internal::checkGlobalErrorStateOrClear( U, ret );
```

Comment:
An alternative way of implementing triu might be to call tril with a `transpose` descriptor?

Reply:
Already the case if you have a look at the _referenec_ implementation: `RC trilu_generic(...)`

Reply:
The thing is that we have to be very clear on what we designate with the word `transpose`:
- If we talk about the input matrix **only**, then `tril` is not equivalent to `triu<transpose>`
e.g 
```
<<< tril<transpose_matrix>(A) >>>
=== MATRIX ===
Size: 10 x 10
0  0  0  0  0  0  0  0  0  0
0  0  0  0  0  0  0  0  0  0
0  0  0  0  0  0  0  0  0  0
3 0  0  9 0  0  0  0  0  0
0  0  0  0  0  0  0  0  0  0
0  0  0  11 13 0  0  0  0  0
6 0  0  0  0  0  18 0  0  0
0  9 0  0  0  0  0  0  0  0
0  0  0  0  0  0  0  0  0  0
0  0  0  0  0  0  0  23 0  0
==============

<<< triu(A) >>>
=== MATRIX ===
Size: 10 x 10
0  0  0  3 0  0  6 0  0  0
0  0  0  0  0  0  0  9 0  0
0  0  0  0  0  0  0  0  0  0
0  0  0  9 0  11 0  0  0  0
0  0  0  0  0  13 0  0  0  0
0  0  0  0  0  0  0  0  0  0
0  0  0  0  0  0  18 0  0  0
0  0  0  0  0  0  0  0  0  23
0  0  0  0  0  0  0  0  0  0
0  0  0  0  0  0  0  0  0  0
==============
```

- If we talk about the input and the output matrix, then `tril` **is** equivalent to `triu<transpose>`. I am not exactly sure how to formally describe such logic using the available descriptors

Reply:
Forget the comment above I was only considering square matrices, plus it's a twisted logic that is absolutely not necessary